wasm_memsafety_rtlib.c

```diff
@@ -64,66 +70,183 @@ static void __wasm_memsafety_table_resize(HashTable *table, size_t capacity) {
+// bytes each and returns a pointer to the allocated memory. The memory is set to
+// zero. If nmemb or size is 0, then calloc() returns either NULL, or a unique
+// pointer value that can later be successfully passed to free().
+void *__wasm_memsafety_calloc(size_t nmemb, size_t size) {
+    if (check_multiplication_overflow(nmemb, size)) {
+        return NULL;
+    }
+    size_t requested_size = nmemb * size;
+
+    // If we don't early exit here on requested size 0, we would allocate
+    // 16 bytes unnecessarily.
+    if (requested_size == 0) {
+        return NULL;
+    }
+
+    void *mem = __wasm_memsafety_malloc(MTE_ALIGNMENT, requested_size);
+    if (mem) {
+        // We only set the requested number of bytes to 0, not any extra bytes
+        // that malloc might have allocated due to alignment.
+        memset(mem, 0, requested_size);
+    }
+
+    return mem;
+}
+
+static inline size_t min(size_t a, size_t b) {
+    return a < b ? a : b;
+}
+
+// TODO: what should happen if we call realloc on a ptr that was previously called with aligned_alloc? Do we have to follow that original alignment?
+// TODO: we're not really setting errno's correctly, should we do that?
+
+// The realloc() function changes the size of the memory block pointed to by ptr
+// to size bytes. The contents will be unchanged in the range from the start of
+// the region up to the minimum of the old and new sizes. If the new size is larger
+// than the old size, the added memory will not be initialized. If ptr is NULL,
+// then the call is equivalent to malloc(size), for all values of size; if size is
+// equal to zero, and ptr is not NULL, then the call is equivalent to free(ptr).
+// Unless ptr is NULL, it must have been returned by an earlier call to malloc(),
+// calloc() or realloc(). If the area pointed to was moved, a free(ptr) is done.
+void *__wasm_memsafety_realloc(void *ptr, size_t requested_size) {
+    if (ptr == NULL) {
+        return __wasm_memsafety_malloc(MTE_ALIGNMENT, requested_size);
+    }
+
+    if (requested_size == 0) {
+        __wasm_memsafety_free(ptr);
+        return NULL;
+    }
+    
+    TableEntry *entry = __wasm_memsafety_table_find(table.entries, table.capacity, ptr);
+    if (entry->key == NULL) {
+        // The pointer wasn't found in the table, indicating the problem that
+        // the ptr was never created with a previous malloc, calloc or realloc.
+        return NULL;
+    }
+
+    // TODO: this assumes an alignment of 16 bytes, but that is not guaranteed (but changing alignment should be allowed since it's undefined in C standard)
+    // If the requested_size is in the last MTE_ALIGNMENT-wide block that was allocated
+    if ((entry->allocated_size - MTE_ALIGNMENT) <= requested_size && requested_size <= entry->allocated_size) {
+        // Just update requested_size
+        __wasm_memsafety_table_set(&table, ptr, requested_size, entry->allocated_size);
```

Comment:
I think in this case we should untag the bytes `[requested_size, allocated_size)` to catch out of bounds errors after realloc, even though nothing would happen for an out of bounds error.

Reply:
I think you might mean another part of the code. Here, we are handling the case where the requested size is inside `[allocated_size - 16, allocated_size)`. This means we don't need to allocate any further memory, but it also means (re)allocating less memory doesn't make sense, because we need the 16-byte alignment anyways. Your proposal to untag the unused part of those 16 bytes would be ideal, but is not possible, since tagging is at a 16-byte granularity itself.

Reply:
Ah right, yes.